src/libpriv/rpmostree-core.c

```diff
@@ -2393,12 +2424,151 @@ canonicalize_non_usrmove_path (RpmOstreeContext  *self,
+      guint off = rpmdbGetIteratorOffset (it);
+      if (g_hash_table_contains (pkgs_deleted, GUINT_TO_POINTER (off)))
+        continue;
+
+      /* try to only load what we need: filenames and colors */
+      rpmfiFlags flags = RPMFI_FLAGS_QUERY;
+#ifdef BUILDOPT_HAVE_RPMFILES /* we're using this more as a "if not on CentOS" switch */
+      /* this is even more restrictive that QUERY */
+      flags = RPMFI_FLAGS_ONLY_FILENAMES;
+#endif
+      flags &= ~RPMFI_NOFILECOLORS;
+
+      g_auto(rpmfi) fi = rpmfiNew (ts, h, RPMTAG_BASENAMES, flags);
+
+      fi = rpmfiInit (fi, 0);
+      while (rpmfiNext (fi) >= 0)
+        {
+          const char *fn = rpmfiFN (fi);
+          g_assert (fn != NULL);
+
+          /* check if one of the pkgs to delete wants to delete our file */
+          if (g_hash_table_contains (files_deleted, fn))
+            g_hash_table_add (files_skip_delete, g_strdup (fn));
+
+          rpm_color_t color = (rpmfiFColor (fi) & ts_color);
+
+          /* let's make the safe assumption that the color mess is only an issue for /usr */
+          const char *fn_rel = fn;
+          fn_rel += strspn (fn, "/");
+
+          /* be sure we've canonicalized usr/ */
+          g_autofree char *fn_rel_owned = canonicalize_non_usrmove_path (self, fn_rel);
+          if (fn_rel_owned)
+            fn_rel = fn_rel_owned;
+
+          if (!g_str_has_prefix (fn_rel, "usr/"))
+            continue;
+
+          /* check if one of the pkgs to install wants to overwrite our file */
+          rpm_color_t other_color =
+            GPOINTER_TO_UINT (g_hash_table_lookup (files_added, fn));
+          other_color &= ts_color;
+
+          /* see handleColorConflict() */
+          if (color && other_color && (color != other_color))
+            {
+              /* do we already have the preferred color installed? */
+              if (color & ts_prefcolor)
+                g_hash_table_add (files_skip_add, canonicalize_rpmfi_path (fn));
+              else if (other_color & ts_prefcolor)
+                {
+                  /* the new pkg is bringing our favourite color, give way now so we let
+                   * checkout silently write into it */
+                  if (!glnx_shutil_rm_rf_at (tmprootfs_dfd, fn_rel, cancellable, error))
+                    return FALSE;
+                }
+            }
+        }
+    }
+
+  *out_files_skip_add = g_steal_pointer (&files_skip_add);
+  *out_files_skip_delete = g_steal_pointer (&files_skip_delete);
+  return TRUE;
+}
+
+static OstreeRepoCheckoutFilterResult
+checkout_filter (OstreeRepo         *self,
+                 const char         *path,
+                 struct stat        *st_buf,
+                 gpointer            user_data)
+{
+  GHashTable *files_skip = user_data;
+  if (g_hash_table_contains (files_skip, path))
+    return OSTREE_REPO_CHECKOUT_FILTER_SKIP;
+  /* hack for nsswitch.conf, which we modify during treecompose; a better heuristic here
```

Comment:
I don't understand why this is necessary now; what would break without this?

Reply:
I expanded on the comment a bit to clarify.